src/core/components.py

```diff
@@ -793,3 +817,138 @@ def get_base_component_dict(self):
         yield from self.components
+
+
+class Panel(Component):
+    """A panel which can contain components."""
+
+    __slots__ = [
+        "components",
+        "width",
+        "is_card",
+        "style",
+    ]
+
+    def __init__(
+        self,
+        components,
+        width=12,
+        is_card=False,
+        style="default",
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.components = components
+        self.width = validate_size(width)
+        self.is_card = is_card
+        self.style = validate_style(style)
+
+    def get_base_component_dict(self):
+        return {
+            "type": "panel",
+            "width": self.width,
+            "is_card": self.is_card,
+            "style": self.style,
+            "components": [
+                [component.get_flow_component_dict() for component in row]
+                for row in self.components
+            ],
+        }
+
+    def get_components(self):
+        yield from super().get_components()
+        yield from self.components
+
+    def get_validators(self):
+        yield from super().get_validators()
+
+class PanelRow(Component):
+    """A row which contains components."""
+
+    __slots__ = [
+        "components",
+        "height",
+    ]
+
+    def __init__(self, components, height=12, **kwargs):
+        super().__init__(**kwargs)
+        self.components = self.validate_components(components)
+        self.height = validate_size(height)
+
+    def validate_components(self, components):
+        """Ensure only Panel components are added to PanelRows"""
+        if any(not isinstance(component, Panel) for component in components):
+            raise InvalidArguments("Only Panel components can be added to Panel Rows.")
+        return components
+
+    def get_base_component_dict(self):
+        return {
+            "type": "panel_row",
+            "height": self.height,
+            "components": [
+                [component.get_flow_component_dict() for component in row]
+                for row in self.components
+            ],
+        }
+
+    def get_components(self):
+        """Include components of child components if applicable (e.g. child panel components)"""
+        yield from super().get_components()
+        yield from self.components
+        yield from [component.get_components() for component in self.components]
+
+
+class Spacer(Component):
+    """A component used for spacing."""
+
+    __slots__ = ["amount"]
+
+    def __init__(self, amount=1, **kwargs):
+        super().__init__(**kwargs)
+        self.amount = self.validate_amount(amount)
+
+    def validate_amount(self, amount):
+        if amount < 1:
+            raise InvalidArguments("Amount cannot be below 1.")
+        return amount
+
+    def get_base_component_dict(self):
+        return {
+            "type": "spacer",
+            "amount": self.amount,
+        }
+
+
+class InstructionBox(Component):
```

Comment:
Isn't this a message box with a style?

Reply:
The idea is it will display a message in a box, but then a direction can be optionally set. E.g. a box saying "Go left" with an arrow displayed off the left side of the box.

Reply:
Do you have a image or something for this? I still don't understand what you mean by "arrow displayed off the left side of the box."

Reply:
I can't show you the image we're using, but it is something like this (for a down arrow):

![image](https://user-images.githubusercontent.com/7999677/119464129-59f65200-bd3a-11eb-8e76-f04ae7b88865.png)

Reply:
That definitely seems like a style then, `style=["primary", "left-arrow-boarder"]` or something like that. Then it's just a text box :) With workflows we should try to have the minimal number of components for no other reason that it minimizes the number of things which need to implimented on the other end,